File: typescript/src/modernimage.ts
import * as fs from 'fs'
import * as os from 'os'
import * as path from 'path'
import {
  modernimage_context_new,
  modernimage_context_free,
  modernimage_set_stdin,
  modernimage_cwebp,
  modernimage_gif2webp,
  modernimage_avifenc,
  modernimage_jpegtran,
  modernimage_jpeg_orientation,
  modernimage_get_exit_code,
  modernimage_get_stderr_size,
  modernimage_copy_stderr,
  modernimage_version,
} from './ffi'
import { ModernImageError, EncodeResult } from './types'
import { cpus } from 'os'

function detectFormat(data: Buffer): string {
  if (data.length < 4) return ''
  if (data[0] === 0xff && data[1] === 0xd8 && data[2] === 0xff) return 'jpeg'
  if (data[0] === 0x89 && data[1] === 0x50 && data[2] === 0x4e && data[3] === 0x47) return 'png'
  if (data[0] === 0x47 && data[1] === 0x49 && data[2] === 0x46) return 'gif'
  return ''
}

let tmpCounter = 0
function createTempPath(suffix: string): string {
  tmpCounter++
  return path.join(os.tmpdir(), `modernimage-${process.pid}-${Date.now()}-${tmpCounter}${suffix}`)
}

type ToolFunc = (ctx: any, argc: number, argv: string[]) => number

function callTool(
  toolFunc: ToolFunc,
  inputData: Buffer,
  argv: string[],
  useStdin: boolean,
): Buffer {
  const ctx = modernimage_context_new()
  if (!ctx) {
    throw new ModernImageError('failed to create context')
  }

  try {
    if (useStdin) {
      modernimage_set_stdin(ctx, inputData, inputData.length)
    }

    const rc = toolFunc(ctx, argv.length, argv)

    if (rc !== 0) {
      const stderrSize = modernimage_get_stderr_size(ctx)
      let errMsg = `tool exited with code ${rc}`
      if (stderrSize > 0) {
        const errBuf = Buffer.alloc(Number(stderrSize))
        modernimage_copy_stderr(ctx, errBuf, errBuf.length)
        errMsg += `: ${errBuf.toString('utf8')}`
      }
      throw new ModernImageError(errMsg)
    }

    // Find output path from argv (-o or -outfile <path>)
    let outPath = ''
    for (let i = 0; i < argv.length; i++) {
      if ((argv[i] === '-o' || argv[i] === '-outfile') && i + 1 < argv.length) {
        outPath = argv[i + 1]
        break
      }
    }

    if (!outPath) {
      throw new ModernImageError('no output path in argv')
    }

    const outData = fs.readFileSync(outPath)
    if (outData.length === 0) {
      throw new ModernImageError('encoding produced empty output')
    }

    return outData
  } finally {
    modernimage_context_free(ctx)
  }
}

/**
 * Encode JPEG/PNG to lossy WebP.
 * ICC profiles are preserved from the source image.
 */
export function encodeLossy(data: Buffer, quality: number = 80, multithread: boolean = false): EncodeResult {
  if (!data || data.length === 0) throw new ModernImageError('empty input data')
  const format = detectFormat(data)
  if (format !== 'jpeg' && format !== 'png') {
    throw new ModernImageError(`unsupported format for encodeLossy (expected JPEG or PNG, got "${format}")`)
  }

  const tmpOut = createTempPath('.webp')
  try {
    const argv = ['cwebp', '-q', String(quality), '-metadata', 'icc']
    if (multithread) argv.push('-mt')
    argv.push('-o', tmpOut, '--', '-')

    const outData = callTool(modernimage_cwebp, data, argv, true)
    return { data: outData, mimeType: 'image/webp' }
  } finally {
    try { fs.unlinkSync(tmpOut) } catch {}
  }
}

/**
 * Encode JPEG/PNG to lossless WebP.
 * ICC profiles are preserved from the source image.
 */
export function encodeLossless(data: Buffer, multithread: boolean = false): EncodeResult {
  if (!data || data.length === 0) throw new ModernImageError('empty input data')
  const format = detectFormat(data)
  if (format !== 'jpeg' && format !== 'png') {
    throw new ModernImageError(`unsupported format for encodeLossless (expected JPEG or PNG, got "${format}")`)
  }

  const tmpOut = createTempPath('.webp')
  try {
    const argv = ['cwebp', '-lossless', '-metadata', 'icc']
    if (multithread) argv.push('-mt')
    argv.push('-o', tmpOut, '--', '-')

    const outData = callTool(modernimage_cwebp, data, argv, true)
    return { data: outData, mimeType: 'image/webp' }
  } finally {
    try { fs.unlinkSync(tmpOut) } catch {}
  }
}

/**
 * Encode GIF to animated WebP.
 */
export function encodeGif(data: Buffer, multithread: boolean = false): EncodeResult {
  if (!data || data.length === 0) throw new ModernImageError('empty input data')
  const format = detectFormat(data)
  if (format !== 'gif') {
    throw new ModernImageError(`unsupported format for encodeGif (expected GIF, got "${format}")`)
  }

  const tmpIn = createTempPath('.gif')
  const tmpOut = createTempPath('.webp')
  try {
    fs.writeFileSync(tmpIn, data)

    const argv = ['gif2webp']
    if (multithread) argv.push('-mt')
    argv.push(tmpIn, '-o', tmpOut)

    const outData = callTool(modernimage_gif2webp, data, argv, false)
    return { data: outData, mimeType: 'image/webp' }
  } finally {
    try { fs.unlinkSync(tmpIn) } catch {}
    try { fs.unlinkSync(tmpOut) } catch {}
  }
}

// AVIF preset parameters matching libnextimage-lite's avifenc_bridge.
interface AvifPreset {
  speed: number
  defaultJobs: number  // 0 = CPU count
  yuv: string          // "444" or "420"
  cicp: string         // "CP/TC/MC"
  depth?: number       // bit depth (omit = default 8)
  autoTile: boolean
  tileRows?: number    // tilerowslog2
  tileCols?: number    // tilecolslog2
}

const presetBalanced: AvifPreset = {
  speed: 6, defaultJobs: 16, yuv: '444', cicp: '1/1/1', autoTile: true,
}
const presetCompact: AvifPreset = {
  speed: 0, defaultJobs: 0, yuv: '444', cicp: '1/1/1', depth: 10, autoTile: true,
}
const presetFast: AvifPreset = {
  speed: 9, defaultJobs: 16, yuv: '420', cicp: '1/13/6',
  autoTile: false, tileRows: 6, tileCols: 6,
}

function encodeAvif(data: Buffer, quality: number, jobs: number, preset: AvifPreset, opName: string): EncodeResult {
  if (!data || data.length === 0) throw new ModernImageError('empty input data')
  const format = detectFormat(data)
  if (format !== 'jpeg' && format !== 'png') {
    throw new ModernImageError(`unsupported format for ${opName} (expected JPEG or PNG, got "${format}")`)
  }

  const tmpOut = createTempPath('.avif')
  try {
    const argv = [
      'avifenc',
      '-q', String(quality),
      '-s', String(preset.speed),
      '--yuv', preset.yuv,
      '--cicp', preset.cicp,
    ]

    if (preset.depth) argv.push('-d', String(preset.depth))

    let effectiveJobs = preset.defaultJobs
    if (jobs > 0) effectiveJobs = jobs
    else if (effectiveJobs === 0) effectiveJobs = cpus().length
    argv.push('-j', String(effectiveJobs))

    if (preset.autoTile) argv.push('--autotiling')
    if (preset.tileRows) argv.push('--tilerowslog2', String(preset.tileRows))
    if (preset.tileCols) argv.push('--tilecolslog2', String(preset.tileCols))

    argv.push('-a', 'tune=ssimulacra2')
    argv.push('--input-format', format, '-o', tmpOut, '--stdin')

    const outData = callTool(modernimage_avifenc, data, argv, true)
    return { data: outData, mimeType: 'image/avif' }
  } finally {
    try { fs.unlinkSync(tmpOut) } catch {}
  }
}

/**
 * Encode JPEG/PNG to AVIF with balanced speed/quality.
 * YUV444, CICP 1/1/1 (BT.709), speed 6, autotiling, tune=ssimulacra2.
 */
export function encodeBalanced(data: Buffer, quality: number = 80, jobs: number = 0): EncodeResult {
  return encodeAvif(data, quality, jobs, presetBalanced, 'encodeBalanced')
}

/**
 * Encode JPEG/PNG to AVIF with best compression (slowest).
 * YUV444, CICP 1/1/1 (BT.709), speed 0, 10-bit depth, autotiling, tune=ssimulacra2.
 */
export function encodeCompact(data: Buffer, quality: number = 80, jobs: number = 0): EncodeResult {
  return encodeAvif(data, quality, jobs, presetCompact, 'encodeCompact')
}

/**
 * Encode JPEG/PNG to AVIF with fastest speed.
 * YUV420, CICP 1/13/6 (BT.709/sRGB/BT.601), speed 9, 64x64 tiling, tune=ssimulacra2.
 */
export function encodeFast(data: Buffer, quality: number = 80, jobs: number = 0): EncodeResult {
  return encodeAvif(data, quality, jobs, presetFast, 'encodeFast')
}

/**
 * Returns the EXIF orientation value (1-8) from JPEG data.
 * Returns 0 if no orientation tag is found or data is not JPEG.
 * This is a pure read-only operation — very fast, no decompression needed.
 */
export function jpegOrientation(data: Buffer): number {
  if (!data || data.length === 0) return 0
  return modernimage_jpeg_orientation(data, data.length)
}

/**
 * Applies lossless rotation based on EXIF orientation, then strips EXIF
 * metadata (preserving ICC profile).
 * If orientation is 1 (normal) or not present, returns the original data unchanged.
 */
export function normalizeJpegOrientation(data: Buffer): Buffer {
  if (!data || data.length === 0) throw new ModernImageError('empty input data')

  const ori = jpegOrientation(data)
  if (ori <= 1) return data

  const transformArgs: string[] = (() => {
    switch (ori) {
      case 2: return ['-flip', 'horizontal']
      case 3: return ['-rotate', '180']
      case 4: return ['-flip', 'vertical']
      case 5: return ['-transpose']
      case 6: return ['-rotate', '90']
      case 7: return ['-transverse']
      case 8: return ['-rotate', '270']
      default: return []
    }
  })()

  if (transformArgs.length === 0) return data

  const tmpOut = createTempPath('.jpg')
  try {
    const argv = ['jpegtran', '-copy', 'icc', '-trim', ...transformArgs, '-outfile', tmpOut]
    return callTool(modernimage_jpegtran, data, argv, true)
  } finally {
    try { fs.unlinkSync(tmpOut) } catch {}
  }
}

/**
 * Get the libmodernimage version string.
 */
export function version(): string {
  return modernimage_version()
}

File: rust/src/bin/modernimage.rs
use clap::{Parser, Subcommand};
use modernimage::{avif, version, webp};
use std::fs;
use std::io::{self, Read, Write};
use std::process;

#[derive(Parser)]
#[command(
    name = "modernimage",
    version = option_env!("MODERNIMAGE_VERSION").unwrap_or(env!("CARGO_PKG_VERSION")),
    about = "WebP and AVIF image encoding"
)]
struct Cli {
    #[command(subcommand)]
    command: Commands,
}

#[derive(Subcommand)]
enum Commands {
    /// WebP encoding commands
    Webp {
        #[command(subcommand)]
        command: WebpCommands,
    },
    /// AVIF encoding commands
    Avif {
        #[command(subcommand)]
        command: AvifCommands,
    },
    /// Show library version
    Version,
}

#[derive(Subcommand)]
enum WebpCommands {
    /// Encode to lossy WebP
    EncodeLossy {
        /// Input file (use - for stdin)
        input: String,
        /// Quality 0-100
        #[arg(short, long, default_value = "80")]
        quality: u32,
        /// Enable multi-threaded encoding
        #[arg(short, long)]
        multithread: bool,
        /// Output file (default: stdout)
        #[arg(short, long)]
        output: Option<String>,
    },
    /// Encode to lossless WebP
    EncodeLossless {
        /// Input file (use - for stdin)
        input: String,
        /// Enable multi-threaded encoding
        #[arg(short, long)]
        multithread: bool,
        /// Output file (default: stdout)
        #[arg(short, long)]
        output: Option<String>,
    },
    /// Encode GIF to animated WebP
    EncodeGif {
        /// Input file (use - for stdin)
        input: String,
        /// Enable multi-threaded encoding
        #[arg(short, long)]
        multithread: bool,
        /// Output file (default: stdout)
        #[arg(short, long)]
        output: Option<String>,
    },
}

#[derive(Subcommand)]
enum AvifCommands {
    /// Encode to AVIF (balanced speed/quality)
    EncodeBalanced {
        /// Input file (use - for stdin)
        input: String,
        /// Quality 0-100
        #[arg(short, long, default_value = "80")]
        quality: u32,
        /// Thread count (0 = auto)
        #[arg(short, long, default_value = "0")]
        jobs: u32,
        /// Output file (default: stdout)
        #[arg(short, long)]
        output: Option<String>,
    },
    /// Encode to AVIF (best compression, slowest)
    EncodeCompact {
        /// Input file (use - for stdin)
        input: String,
        /// Quality 0-100
        #[arg(short, long, default_value = "80")]
        quality: u32,
        /// Thread count (0 = auto)
        #[arg(short, long, default_value = "0")]
        jobs: u32,
        /// Output file (default: stdout)
        #[arg(short, long)]
        output: Option<String>,
    },
    /// Encode to AVIF (fastest)
    EncodeFast {
        /// Input file (use - for stdin)
        input: String,
        /// Quality 0-100
        #[arg(short, long, default_value = "80")]
        quality: u32,
        /// Thread count (0 = auto)
        #[arg(short, long, default_value = "0")]
        jobs: u32,
        /// Output file (default: stdout)
        #[arg(short, long)]
        output: Option<String>,
    },
}

fn read_input(path: &str) -> Vec<u8> {
    if path == "-" {
        let mut buf = Vec::new();
        io::stdin().read_to_end(&mut buf).unwrap_or_else(|e| {
            eprintln!("Error reading stdin: {}", e);
            process::exit(1);
        });
        buf
    } else {
        fs::read(path).unwrap_or_else(|e| {
            eprintln!("Error reading {}: {}", path, e);
            process::exit(1);
        })
    }
}

fn write_output(data: &[u8], output: &Option<String>) {
    let result = match output {
        Some(path) => fs::write(path, data),
        None => io::stdout().write_all(data),
    };
    if let Err(e) = result {
        eprintln!("Error writing output: {}", e);
        process::exit(1);
    }
}

fn run_encode<F>(input: &str, output: &Option<String>, encode_fn: F)
where
    F: FnOnce(&[u8]) -> Result<modernimage::EncodeResult, modernimage::error::ModernImageError>,
{
    let data = read_input(input);
    match encode_fn(&data) {
        Ok(r) => write_output(&r.data, output),
        Err(e) => {
            eprintln!("Error: {}", e);
            process::exit(1);
        }
    }
}

fn main() {
    let cli = Cli::parse();

    match cli.command {
        Commands::Webp { command } => match command {
            WebpCommands::EncodeLossy {
                input,
                quality,
                multithread,
                output,
            } => run_encode(&input, &output, |d| webp::encode_lossy(d, quality, multithread)),

            WebpCommands::EncodeLossless {
                input,
                multithread,
                output,
            } => run_encode(&input, &output, |d| webp::encode_lossless(d, multithread)),

            WebpCommands::EncodeGif {
                input,
                multithread,
                output,
            } => run_encode(&input, &output, |d| webp::encode_gif(d, multithread)),
        },

        Commands::Avif { command } => match command {
            AvifCommands::EncodeBalanced {
                input,
                quality,
                jobs,
                output,
            } => run_encode(&input, &output, |d| avif::encode_balanced(d, quality, jobs)),

            AvifCommands::EncodeCompact {
                input,
                quality,
                jobs,
                output,
            } => run_encode(&input, &output, |d| avif::encode_compact(d, quality, jobs)),

            AvifCommands::EncodeFast {
                input,
                quality,
                jobs,
                output,
            } => run_encode(&input, &output, |d| avif::encode_fast(d, quality, jobs)),
        },

        Commands::Version => {
            println!("modernimage CLI");
            println!("libmodernimage version: {}", version());
        }
    }
}
